### PR TITLE
Bug - Azure Sentinel fetch incidents - incidents do not get fetched by xsoar

### PR DESCRIPTION
**Pull request by MLainer1**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/46408

## Description
A bug where  incidents are not digested by fetch-incidents command.
Possible cause:
image
incident 553824 has a Created Time=03:32, but incident 553823 Created Time=03:33, and 553824 did not get fetched, because we are stating the next fetch from the last incident Created Time.
In this example-> incident 553823 Created Time=03:33, so we started fetching incidents that were created from 03:33 and on. 553824 has a Created Time=03:32, so even though it has a higher ID, we did not fetched it because of its created time.
You can see it again in incidents 553838, 553837.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py [hunk 1]
```diff
@@ -457,27 +457,34 @@ def test_module(client):
...
 
-def list_incidents_command(client, args, is_fetch_incidents=False):
+def list_incidents_command(client, args, is_fetch_incidents=False, first_fetch=False):
+    demisto.debug("starting the list incidents command")
     filter_expression = args.get('filter')
     limit = None if is_fetch_incidents else min(200, int(args.get('limit')))
     next_link = args.get('next_link', '')
+    result = None
+    params = {}
+    demisto.debug(f"is_fetch_incidents:{is_fetch_incidents}, first_fetch:{first_fetch}")
```
Comment (jochman) [inline Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py:467]:
```suggestion
    demisto.debug(f"{is_fetch_incidents=}, {first_fetch=}")
```
Sweet sweet python 3.8 syntax sugaring

File: Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py [hunk 2]
```diff
@@ -868,27 +875,53 @@ def fetch_incidents(client, last_run, first_fetch_time, min_severity):
     # Get the last fetch details, if exist
     last_fetch_time = last_run.get('last_fetch_time')
     last_fetch_ids = last_run.get('last_fetch_ids', [])
+    last_incident_number = last_run.get('last_incident_number')
+    demisto.debug(f"last fetch time is: {last_fetch_time}, last fetch ids are:{last_fetch_ids}")
```
Comment (jochman) [inline Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py:879]:
```suggestion
    demisto.debug(f"{last_fetch_time=}, {last_fetch_ids=}")
```
Sweet sweet python 3.8 syntax sugaring

File: Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py [hunk 3]
```diff
@@ -868,27 +875,53 @@ def fetch_incidents(client, last_run, first_fetch_time, min_severity):
...
 
-    for incident in items:
+    for incident in raw_incidents:
         incident_severity = severity_to_level(incident.get('Severity'))
+        demisto.debug(f"incident id: {incident.get('ID')}, severity level:{incident_severity}, incident number:{incident.get('IncidentNumber')}")
 
         # fetch only incidents that weren't fetched in the last run and their severity is at least min_severity
-        if incident.get('ID') not in last_fetch_ids and incident_severity >= min_severity:
+        # and their id is larger then the previous incident
+        if incident.get('ID') not in last_fetch_ids and incident_severity >= min_severity and last_incident_number != incident.get('IncidentNumber'):
```
Comment (jochman) [inline Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py:923]: I think we might want to break it down into two functions -> by id order and by time order. Here we have two different flows combined, and it isn't apparent.

**Review by jochman — COMMENTED** [review]
looks fine. Please note the (redundant) extralogic with ids.

File: Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py [hunk 4]
```diff
@@ -457,23 +459,32 @@ def test_module(client):
...
+
     else:
-        url_suffix = 'incidents'
-        params = {
-            '$top': limit,
-            '$filter': filter_expression,
-            '$orderby': 'properties/createdTimeUtc asc'
-        }
+        params = {'$filter': args.get('$filter', ''),
+                  '$orderby': args.get('$orderby', ''),
```
Comment (dantavori) [inline Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py:482]:
```suggestion
                  '$orderby': args.get('orderby', 'properties/createdTimeUtc asc'),
```

File: Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py [hunk 5]
```diff
@@ -457,23 +459,32 @@ def test_module(client):
...
         result = client.http_request('GET', full_url=next_link)
+
     else:
-        url_suffix = 'incidents'
-        params = {
-            '$top': limit,
-            '$filter': filter_expression,
-            '$orderby': 'properties/createdTimeUtc asc'
-        }
+        params = {'$filter': args.get('$filter', ''),
```
Comment (dantavori) [inline Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py:481]:
```suggestion
        params = {
            '$filter': args.get('filter', ''),
```
please stick to [pep8 style conventions](https://www.python.org/dev/peps/pep-0008/#indentation) (also verify your other changes)

File: Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py [hunk 6]
```diff
@@ -864,31 +875,100 @@ def update_next_link_in_context(result: dict, outputs: dict):
...
+        last_fetch_time: The last fetch time.
+        client: an AzureSentinelClient client.
+
+    Returns:
+        An array of incidents, and a latest created time.
+    """
+
+    latest_created_time_str = last_fetch_time.strftime(DATE_FORMAT)
+    command_args = {'$filter': f'properties/createdTimeUtc ge {latest_created_time_str}',
+                    '$orderby': 'properties/createdTimeUtc asc'}
```
Comment (dantavori) [inline Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py:922]:
```suggestion
    command_args = {'filter': f'properties/createdTimeUtc ge {latest_created_time_str}',
                    'orderby': 'properties/createdTimeUtc asc'}
```

File: Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py [hunk 7]
```diff
@@ -864,31 +875,100 @@ def update_next_link_in_context(result: dict, outputs: dict):
...
+def fetch_incidents(client: AzureSentinelClient, last_run: dict, first_fetch_time: str, min_severity: int):
+    """Fetching incidents.
+    Args:
+        first_fetch_time: The first fetch time.
+        client: An AzureSentinelClient client.
+        last_run: An dictionary of the last run.
+        min_severity: A minimum severity of incidents to fetch.
+
+    Returns:
+        An array of incidents, and a latest created time.
```
Comment (dantavori) [inline Packs/AzureSentinel/Integrations/AzureSentinel/AzureSentinel.py:887]:
```suggestion
        (tuple): 1. The LastRun object updated with the last run details.
                 2. An array of incidents.
```

**Review by dantavori — CHANGES_REQUESTED** [review]
please apply the changes today so we can release this to customers asap

**Review by dantavori — APPROVED** [review]
great, before merging make sure it works properly in your env

**Pull request merged**